doc/src/devdocs/compiler.md

```diff
@@ -0,0 +1,119 @@
+gets recreated and contains the right runtime pointer.
+
+Otherwise, they will be emitted as literal constants.
+
+To emit one of these objects, call `literal_pointer_val`.
+It'll handle tracking the Julia value and the LLVM global,
+ensuring they are valid both for the current runtime and after deserialization.
+
+When emitted into the object file, these globals are stored as references
+in a large `gvals` table. This allows the deserializer to reference them by index,
+and implement a custom manual GOT-like mechanism to restore them.
+
+Function pointers are handled similarly.
+They are stored as values in a large `fvals` table.
+Like globals, this allows the deserializer to reference them by index.
+
+Note that extern functions are handled separately,
+with names, via the usual symbol resolution mechanism in the linker.
+
+Note too that ccall functions are also handled separately,
+via a manual GOT + PLT.
+
+
+## Representation of Intermediate Values
+
+Values are passed around in a `jl_cgval_t` struct.
+This represents an R-value, and includes enough information to
+determine how to assign or pass it somewhere.
+
+They are created via one of the helper constructors, usually:
+`mark_julia_type` (for immediate values) and `mark_julia_slot` (for pointers to values).
+
+The function `convert_julia_type` can transform between any two types.
+When it returns and returns an `cgval.typ` set to `typ`.
+It'll cast the object to the requested representation.
+It'll make boxes, allocate stack copies, and compute tagged unions as
+needed to perform the request.
+
+By contrast `update_julia_type` will change `cgval.typ` to `typ`,
+only if it can be done at zero-cost (i.e. without emitting any code).
+
+
+## Union representation
+
+Inferred union types may be stack allocated via a tagged type representation.
+
+The primitive routines that need to be able to handle tagged unions are:
+- remark-type
+- load-local
+- store-local
+- isa
+- is
+- emit_typeof
+- emit_sizeof
+- boxed
+- unbox
+- specialized cc-ret
+
+Everything else should be possible to handle in inference by using these
+primitives to implement union-splitting.
+
+The representation of the tagged-union is as a pair
+of < void* union, byte selector >.
+The selector is fixed-size as `byte & 0x7f`,
+and will union-tag the first 126 isbits.
+It records the one-based depth-first count into the type-union of the
+isbits objects inside. An index of zero indicates that the `union*` is
+actually a tagged heap-allocated `jl_value_t*`,
+and needs to treated as normal for a boxed object rather than as a
+tagged union.
+
+The high bit of the selector (`byte & 0x80`) can be tested to determine if the
+`void*` if the `void*` is actually a heap-allocated box,
+thus avoiding the cost of re-allocating a box,
+while maintaining the ability to efficiently handle union-splitting based on the low bits.
+
+It is guaranteed that `byte & 0x7f` is an exact test for the type,
```

Comment:
I think either "since" should follow the comma, or the comma should be replaced by a semicolon.